### ISSUE TITLE
Painless script_fields don't have access to a _source variable

### ISSUE THREAD

**Issue by Bargs** [elastic/elasticsearch#20068]
**Elasticsearch version**: master

**Plugins installed**: none

**JVM version**:  1.8.0_60

**OS version**: OSX El Capitan

**Description of the problem including expected versus actual behavior**:

According to [the docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/search-request-script-fields.html) `script_fields` can access the document _source via a `_source` keyword. When attempting this with a Painless script, I get a `Variable [_source] is not defined.` error.

**Steps to reproduce**:
 1. Execute a query with a painless script_field that attempts to access _source, something like this:

```
    "script_fields": {
        "scriptip": {
            "script": {
                "inline": "return _source.ip;",
                "lang": "painless"
            }
        }
    }
```

**Provide logs (if relevant)**:

```
{
  "error": {
    "root_cause": [
      {
        "type": "script_exception",
        "reason": "compile error",
        "script_stack": [
          "return _source.ip;",
          "       ^---- HERE"
        ],
        "script": "return _source.ip;",
        "lang": "painless"
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query_fetch",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "logstash-0",
        "node": "00s71MPnSRenvz6mflLQeQ",
        "reason": {
          "type": "script_exception",
          "reason": "compile error",
          "caused_by": {
            "type": "illegal_argument_exception",
            "reason": "Variable [_source] is not defined."
          },
          "script_stack": [
            "return _source.ip;",
            "       ^---- HERE"
          ],
          "script": "return _source.ip;",
          "lang": "painless"
        }
      }
    ],
    "caused_by": {
      "type": "script_exception",
      "reason": "compile error",
      "caused_by": {
        "type": "illegal_argument_exception",
        "reason": "Variable [_source] is not defined."
      },
      "script_stack": [
        "return _source.ip;",
        "       ^---- HERE"
      ],
      "script": "return _source.ip;",
      "lang": "painless"
    }
  },
  "status": 500
}
```

**Describe the feature**:

**Comment by jdconrad** [comment]
_source is definitely a controversial issue, but as of right now, _source is only supported in update scripts within Painless.

**Comment by Bargs** [comment]
Sounds like the script fields docs should be updated to at least note that then. Right now they're a bit misleading because the `doc` example is in Painless and the `_source` example doesn't mention the script lang explicitly.

**Comment by jdconrad** [comment]
Yeah, I agree completely that the documentation needs to be improved.  I will be investing some significant effort into this (at least for the Painless side) over the next few weeks.

**Comment by rmuir** [comment]
I dont think it needs to be controversial for script fields? I think the fix is to move them to the `ExecutableScript` api, so they have `_source` and other stuff.

The problem is, they are using the `SearchScript` api today, which should be for the millions-of-documents case

**Comment by clintongormley** [comment]
> I think the fix is to move them to the ExecutableScript api, so they have _source and other stuff.

+1  It'll be a breaking change for existing scripts which currently access `_source.` and will need to change to `ctx._source.`, but I think it's the right move.

**Comment by rmuir** [comment]
Right, "breaking change" is one way to look at it. "broken now" is another :)

**Comment by clintongormley** [comment]
@Bargs @epixa Could you give us some examples of how people use script fields? Do people use them to extract values from the source, or from doc values?

**Comment by Bargs** [comment]
@clintongormley definitely doc values up until now. Since Lucene expressions only support numerical data I don't think there was ever much point in accessing source. In fact, it looks like Kibana throws an error if you try to access _source in an expression script. With painless there might be more reason to use source, but it's hard to say how users will leverage it.

An additional (possible) complication is that Kibana scripted fields are made available on the Visualize tab as well, so the scripts can be used in aggregations. Is it possible to access source in scripts on aggregations?

**Comment by nik9000** [comment]
> Is it possible to access source in scripts on aggregations?

In scripts it is possible. Groovy supports it now and painless doesn't. It is always much, much, much slower than using doc_values. It is a speed vs flexibility tradeoff vs education thing.

**Comment by s1monw** [comment]
I am personally convinced we should never access _source in an aggregation. I think the majority of the users do NOT understand what that means ie.  decompressing the ENTIRE document to access a single field. I don't really think kibana should expose it that way sorry. I am -1 on accessing _source in aggs via painless

**Comment by clintongormley** [comment]
> I am -1 on accessing _source in aggs via painless

Agreed.  Question is whether script fields should be able to access source?  I believe executable scripts don't expose doc values today, while search scripts don't expose source.  Script fields are in no-man's land between the two:  it should have access to both.

**Comment by rjernst** [comment]
Give 5.0 is almost ready, there is an existing workaround I think we can "promote" for now, but know that it will break in the future. In painless, _source is currently available via `params._source`. This is due to a side effect of the existing scripts infrastructure, and was never intended to be used. The long term path forward is to make `_source` be a local variable available depending on the context used (for example, it would not be available in aggs or search scripts, but would be there for script fields, see #20621).

Given this workaround, I'm removing the breaking label from here. For 5.x, when backporting script context work, we can add deprecation warnings if using `params._source`.

**Comment by seschulz** [comment]
I would like to add that the script parameter for significant terms aggregation can not be accessed directly by painless scripts as well (like _subset_freq or _superset_freq).

The example given in the documentation is therefore incorrect.
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-significantterms-aggregation.html#_scripted

The mentioned workaround with the "params." prefix still works for me.

**Comment by clintongormley** [comment]
@seschulz that's not  a workaround, but the correct way to access those params.  I've updated the example in the docs accordingly.  thanks for pointing it out

**Comment by dbason** [comment]
FWIW our use case for accessing source is to use the message field from Logstash indices with a regex to create an extra field.  We do this in Kibana with scripted fields and also use it in aggregations (via groovy scripts) for use with Watcher.

Admittedly this is just a stop gap until we get the field added in to Logstash correctly, but a useful one none the less.

**Comment by nezda** [comment]
`script_fields` docs are currently wrong for this too https://www.elastic.co/guide/en/elasticsearch/reference/5.2/search-request-script-fields.html gives example:

```json
{
        "query" : {
            "match_all": {}
        },
        "script_fields" : {
            "test1" : {
                "script" : "_source.obj1.obj2"
            }
        }
    }
```

which doesn't work - should be 
```
{
        "query" : {
            "match_all": {}
        },
        "script_fields" : {
            "test1" : {
                "script" : "params._source.obj1.obj2"
            }
        }
    }
```

**Comment by jdconrad** [comment]
@nezda Thanks for pointing this out.  This should be fixed with #21322.

**Comment by jeacott1** [comment]
I'm trying to sum across fields in an aggregation and am currently doing something like this: 
      "aggs": {
        "1": {
          "sum": {
            "script": "doc['my.field1'].value + doc['my.field2'].value"
          }
        }
      }

is there a better way?

unfortunately not all docs contain my.field2. 
I wanted to simply include an if(exists('my.field2'...
and the only way I've been able to even vaguely manage this is by checking if it exists in the _source. 
its far from ideal, and it doesn't work in any form I can find inside an aggregation. 
I wanted something like this (prefer something that doesnt touch source but idk what that is).
ctx._source.containsKey('my.field2')?doc['my.field2'].value:0

so to the question of usecases here's one.  any hints as to how this can be achieved?

**Comment by clintongormley** [comment]
@jeacott1 this works just fine:

```
PUT t/t/1
{
  "foo": 1
}

PUT t/t/2
{
  "foo": 3,
  "bar": 5
}

GET t/_search?size=0
{
  "aggs": {
    "sum_two_fields": {
      "sum": {
        "script": {
          "inline": "doc['foo'].value + doc['bar'].value"
        }
      }
    }
  }
}
```
returns

```
{
  "took": 4,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "failed": 0
  },
  "hits": {
    "total": 2,
    "max_score": 0,
    "hits": []
  },
  "aggregations": {
    "sum_two_fields": {
      "value": 9
    }
  }
}
```

**Comment by jeacott1** [comment]
@clintongormley thanks for responding, I'm searching across 2 indexes, once contains the metric and one doesn't.
I think I know whats going on. I only get results for the index containing both fields, so I get an exception, plus results for the other index. So I'm missing half the results.

I have a dynamic template setting types like this
```
 "dynamic_templates": [
          {
            "device": {
              "path_match": "dev.int.*",
              "match_mapping_type": "long",
              "mapping": {
                "dynamic": "true",
                "type": "integer"
              }
```
after query:
```
   "failures": [
      {
        "shard": 0,
        "index": "test",
        "node": "qaGJ4qfTTSSDFVLLcJ42eg",
        "reason": {
          "type": "script_exception",
          "reason": "runtime error",
          "script_stack": [
            "org.elasticsearch.search.lookup.LeafDocLookup.get(LeafDocLookup.java:77)",
            "org.elasticsearch.search.lookup.LeafDocLookup.get(LeafDocLookup.java:36)",
            "doc['dev.int.temp2'].value +  doc['dev.int.temp1'].value",
            "    ^---- HERE"
          ],
          "script": "doc['dev.int.temp2'].value +  doc['dev.int.temp1'].value",
          "lang": "painless",
          "caused_by": {
            "type": "illegal_argument_exception",
            "reason": "No field found for [dev.int.temp2] in mapping with types []"
          }
        }
      }
    ]
 ```

I don't want to explicitly define my field mappings here (thats what dynamic templates are for).

if I explicitly add a mapping on the index for temp2 it works fine. 
```
{
  "properties": {
    "dev.int.temp2": {
          "type": "integer"
    }
  }
}
```

for unknown types in situations like this should not ES refer to the dynamic template definition for guidance?

**Comment by clintongormley** [comment]
@jdconrad any ideas?

**Comment by jdconrad** [comment]
@jeacott1 

Looking through the code, doc in Painless should just be a Map with the method containsKey.  So I believe you can do something like the following:

```"(doc.containsKey('dev.int.temp2') ? doc['dev.int.temp2'].value : 0) +  (doc.containsKey('dev.int.temp1') ? doc['dev.int.temp1'].value : 0)```

If that still gives you problems, please let me know and I'll investigate further.

On the dynamic templates, I'm not completely sure how a dynamic template should cause a non-existent field to behave.  It's probably correct the way it is now as a template shouldn't cause a field to exist if it wasn't previously defined.

**Comment by jeacott1** [comment]
thanks jdconrad, I tried ctx._source, I cant believe I never tried doc.containsKey, it works!

as for templates creating fields that dont exist, sure that seems like the correct solution, but for a query/aggregatoin involving multiple indexes where the field may/may not exist there is enough info in the system to be able to derive what a missing field type is and accommodate, even if its just by rewriting the query with doc.containsKey where appropriate. just a suggestion.

**Comment by jdconrad** [comment]
@jeacott1 It is an interesting thought to have that section of the code re-written for dynamic templates, though given our current architecture this isn't possible.  Maybe something cool for the future!

**Comment by shijiebei2009** [comment]
this works for me 
```
GET actionLog/_search?from=0&size=1
{
  "script_fields": {
    "newCBID": {
      "script": "params._source.CBID-10000000"
    }
  },
  "query": {
    "match_all": {}
  }
}
```

**Comment by shijiebei2009** [comment]
@rjernst 
when using script in elasticsearch, why must we use params._source to access _source in GET query, but use ctx._source to access _source in POST query, es v5.5.0

**Comment by jdconrad** [comment]
@shijiebei2009 This has to do with what's available in different script contexts.  The documentation is being worked on to explain the discrepancies, but isn't available yet.

**Comment by azasypkin** [comment]
@jdconrad was it finally documented somewhere?

**Comment by jdconrad** [comment]
@azasypkin https://www.elastic.co/guide/en/elasticsearch/painless/master/painless-contexts.html

**Comment by Zhang21** [comment]
I have a question about compare fields in dsl query like that.

<br>

What should I do? Thanks.

Determination of the value of the field `host.name`:

```
# it doesn't works
script: {
  "source": "params['_source']['host']['name'].value == 'xxx'"
}
```

```
# it doesn't works
script: {
  "source": "params['_source.host.name'].value == 'xxx'"
}
```

```
# it doesn't works
script: {
  "source": "params._source.host.name == 'xxx'"
}
```

```
# it doesn't works
script: {
  "source": "ctx._source.host.name == 'xxx'"
}
```

**Comment by jdconrad** [comment]
@Zhang21 As this is already closed, and isn't a bug/feature request, please move this to a discuss issue. (https://discuss.elastic.co/) This allows it to get more attention and other users will have better access if they encounter your same problem. Thank you in advance.

**Comment by Zhang21** [comment]
@jdconrad  Thank you.

I found it in docs: <>

> It’s important to understand the difference between doc['my_field'].value and params['_source']['my_field']. The first, using the doc keyword, will cause the terms for that field to be loaded to memory (cached), which will result in faster execution, but more memory consumption. Also, the doc[...] notation only allows for simple valued fields (you can’t return a json object from it) and makes sense only for non-analyzed or single term based fields. However, using doc is still the recommended way to access values from the document, if at all possible, because _source must be loaded and parsed every time it’s used. Using _source is very slow.
